internal/plans/objchange/objchange.go

```diff
@@ -484,165 +478,43 @@ func proposedNewNestedType(schema *configschema.Object, prior, config cty.Value)
-				kv, ev := it.Element()
-				elems[kv.AsString()] = setElementCompareValue(&blockType.Block, ev, isConfig)
-			}
 
-			switch {
-			case elementType.HasDynamicTypes():
-				attrs[name] = cty.ObjectVal(elems)
-			default:
-				attrs[name] = cty.MapVal(elems)
-			}
-
-		default:
-			// Should never happen, since the above cases are comprehensive.
-			panic(fmt.Sprintf("unsupported block nesting mode %s", blockType.Nesting))
+		attr := schema.AttributeByPath(path)
+		if attr == nil {
+			// we must be at a nested block rather than an attribute, continue
+			return v, nil
 		}
-	}
-
-	return cty.ObjectVal(attrs)
-}
-
-// setElementCompareValues takes a known, non-null value of a cty.Set type and
-// returns a table -- constructed of two-element arrays -- that maps original
-// set element values to corresponding values that have all of the computed
-// values removed, making them suitable for comparison with values obtained
-// from configuration. The element type of the set must conform to the implied
-// type of the given schema, or this function will panic.
-//
-// In the resulting slice, the zeroth element of each array is the original
-// value and the one-indexed element is the corresponding "compare value".
-//
-// This is intended to help correlate prior elements with configured elements
-// in proposedNewBlock. The result is a heuristic rather than an exact science,
-// since e.g. two separate elements may reduce to the same value through this
-// process. The caller must therefore be ready to deal with duplicates.
-func setElementCompareValuesFromObject(schema *configschema.Object, set cty.Value) [][2]cty.Value {
-	ret := make([][2]cty.Value, 0, set.LengthInt())
-	for it := set.ElementIterator(); it.Next(); {
-		_, ev := it.Element()
-		ret = append(ret, [2]cty.Value{ev, setElementCompareValueFromObject(schema, ev)})
-	}
-	return ret
-}
 
-// setElementCompareValue creates a new value that has all of the same
-// non-computed attribute values as the one given but has all computed
-// attribute values forced to null.
-//
-// The input value must conform to the schema's implied type, and the return
-// value is guaranteed to conform to it.
-func setElementCompareValueFromObject(schema *configschema.Object, v cty.Value) cty.Value {
-	if v.IsNull() || !v.IsKnown() {
-		return v
-	}
-	attrs := map[string]cty.Value{}
-
-	for name, attr := range schema.Attributes {
-		attrV := v.GetAttr(name)
-		switch {
-		case attr.Computed:
-			attrs[name] = cty.NullVal(attr.Type)
-		default:
-			attrs[name] = attrV
+		if attr.Computed {
+			// If this could have been computed, we always return null in order
+			// to compare config and state values.
+			//
+			// The only way to determine if an optional+computed value is
+			// optional or computed is see if it is set in the config, however
+			// for set elements we don't yet know which element value
+			// corresponds to the configuration.
+			return cty.NullVal(v.Type()), nil
```

Comment:
The code just returns null for any computed value with disregard whether it's specified in config or not (in case it's both computed and optional). That means that the [comparison](https://github.com/hashicorp/terraform/blob/7ca9abe463b8ff44f08bd699304e6b8002615b75/internal/plans/objchange/objchange.go#L447) with config returns false and the proposed new state is set to config instead of the prior state (`cmp[0]`)

Reply:
That is intended. This transformation is used to create value to attempt and correlate config with prior state within a set, it is not directly used as a plan value. Because we don't know what the original config value was, we cannot tell if the prior value was computed or not, so we must assume it was computed.

Reply:
It leads to dirty plans for unchanged configurations with sets that have underlying computed and optional attributes, so sets cannot be used in these cases. At least with the current framework implementation.

Reply:
That is correct. If there are no stable attributes which can correlate configuration to prior state, then you are not going to get a very useful proposed plan. Set elements are identified by their value, so if a set element could have all computed attributes there's no way to determine if it could have originated from the given configuration or not (the simpler example is where an optional+computed value being removed from the configuration and Terraform cannot tell if it was originally computed or not, but made worse by being included within a set).

Using computed value at all within a set would be unusual unless you have very specific needs and fully understand all the implications when trying to plan changes to the values.

Reply:
Sorry, extra changes are expected, but this should not cause "dirty plans" when there are _no_ changes. I thought I confirmed tests for the no-changes case, but let me double check on that,

Reply:
The test fails if we make the [optional attribute](https://github.com/hashicorp/terraform/blob/jbardin/optional-computed-nested-objects/internal/plans/objchange/objchange_test.go#L2156) to be both optional and computed - `"optional": {Type: cty.String, Optional: true, Computed: true},`. The [check for equality](https://github.com/hashicorp/terraform/blob/7ca9abe463b8ff44f08bd699304e6b8002615b75/internal/plans/objchange/objchange.go#L447) fails in that case and the config is used while it should be the prior state.

Reply:
@dimuon, yes, that's what I meant above by there being no non-computed attributes which can be used to correlate config and state values. This is just inherent in trying to use optional+computed values within a set, because we cannot always determine whether a value was generated from config or not. There is no correct value to send in this case, and either will result in false positives, but we happen to send the config when it exists (in most cases, I have some `null` edge cases to consider in a followup PR). That does not mean it can't be fixed up by the provider during the plan, but in practice you should carefully consider whether optional+computed is really what you want to use in the first place.

The specific behavior BTW is captured in [this test](https://github.com/hashicorp/terraform/blob/jbardin/optional-computed-nested-objects/internal/plans/objchange/objchange_test.go#L2425).

Reply:
@jbardin , I see your point. IMO the current design means that set works as expected only if it contains no more than one computed and optional underlying attributes. Once set contains 2 or more such attributes, it can lead to "dirty plan" for configurations with no change (e.g. when configuration defines one of these optional and computed fields and doesn't set any value for the second and state contains a computed value the second attribute). Please correct that if my conclusion is wrong.

Regarding fixing the behavior in the provider. It looks like the default way to do this is plan modifiers. But if the resource is big, the provider has to provide plan modifies for every computed field (or a plan modifier for the whole resource) but writing a plan modifier can be tricky for complex resources because plan modifier should understand relationship between different fields to make a correct decision (e.g. whether it should use the current state of fallback to `Unknown`).

I'm not sure whether the implementation can be improved though. Can we skip nullifying computed attributes before comparing them to config but not take computed attribute into comparison if state has some value for it and config doesn't? 

Anyway, @jbardin , thank you for the interesting discussion.

Reply:
@dimuon, IMHO optional+computed could be used sparingly, and in hindsight probably should not have been allowed within set types, but that was decided long ago ;) 

The problem with your suggestion lies in the fact that sets are unordered data types indexed only by the element value itself, so when the value in config and the value in state are not exactly equal, we cannot tell "if state has some value for it and config doesn't", because we don't know if those were logically the same value in the first place.

I do have another idea for a comparison method I'm going test out later (this PR was mainly aimed at fixing errors in the current behavior, not changing the behavior). While it's not a solvable problem, we may be able to do the comparison using the lifecycle rules that a provider must abide by to try and determine which values may share a common lineage. I don't know yet if that will result in plans looking more closely like the common cases providers expect, or of we're going to generate different changes breaking existing providers.

Reply:
FYI most of this PR will be superseded by a later PR that is going to test a new set comparison method which can check for most of these edge cases.